src/data/equiv/algebra.lean

```diff
@@ -190,6 +199,102 @@ protected def discrete_field [discrete_field β] : discrete_field α :=
 end instances
 end equiv
 
+structure add_equiv (α β : Type*) [has_add α] [has_add β] extends α ≃ β :=
+(hom : is_add_hom to_fun)
+
+infix ` ≃+ `:50 := add_equiv
+
+namespace add_equiv
+
+variables [has_add α] [has_add β] [has_add γ]
+
+@[refl] def refl (α : Type) [has_add α] : α ≃+ α :=
+{ hom := λ _ _,rfl,
+  ..equiv.refl _}
+
+@[symm] def symm (h : α ≃+ β) : β ≃+ α :=
+{ hom := λ n₁ n₂, function.injective_of_left_inverse h.left_inv begin
+   rw h.hom, unfold equiv.symm, rw [h.right_inv, h.right_inv, h.right_inv], end
+  ..h.to_equiv.symm}
+
+@[trans] def trans (h1 : α ≃+ β) (h2 : β ≃+ γ) : (α ≃+ γ) :=
+{ hom := λ _ _, is_add_hom.comp h1.hom h2.hom,
+  ..equiv.trans h1.to_equiv h2.to_equiv }
+
+end add_equiv
+
+structure mul_equiv (α β : Type*) [has_mul α] [has_mul β] extends α ≃ β :=
+(hom : is_mul_hom to_fun)
+
+infix ` ≃* `:50 := mul_equiv
+
+namespace mul_equiv
+
+variables [has_mul α] [has_mul β] [has_mul γ]
+
+@[refl] def refl (α : Type) [has_mul α] : α ≃* α :=
+{ hom := λ _ _,rfl,
+..equiv.refl _}
+
+@[symm] def symm (h : α ≃* β) : β ≃* α :=
+{ hom := λ n₁ n₂, function.injective_of_left_inverse h.left_inv begin
+   rw h.hom, unfold equiv.symm, rw [h.right_inv, h.right_inv, h.right_inv], end
+  ..h.to_equiv.symm}
+
+@[trans] def trans (h1 : α ≃* β) (h2 : β ≃* γ) : (α ≃* γ) :=
+{ hom := λ _ _, is_mul_hom.comp h1.hom h2.hom,
+  ..equiv.trans h1.to_equiv h2.to_equiv }
+
+end mul_equiv
+
+-- equiv of monoids
+namespace mul_equiv
+
+variables [monoid α] [monoid β] [monoid γ]
+
+lemma one (h : equiv α β) (hom : ∀ x y, h (x * y) = h x * h y) :
+  h 1 = 1 :=
+by rw [←mul_one (h 1), ←h.apply_symm_apply 1, ←hom]; simp
+
+instance is_monoid_hom (h : α ≃* β) : is_monoid_hom h.to_equiv := {
+  map_one := mul_equiv.one h.to_equiv h.hom,
+  map_mul := h.hom }
+
+end mul_equiv
+
+-- equiv of groups
+namespace mul_equiv
+
+variables [group α] [group β] [group γ]
+
+instance is_group_hom (h : α ≃* β) : is_group_hom h.to_equiv := ⟨h.hom⟩
+
+end mul_equiv
+
+-- equiv of add_groups
+namespace add_equiv
+
+variables [add_group α] [add_group β] [add_group γ]
+
+instance is_add_group_hom (h : α ≃+ β) : is_add_group_hom h.to_equiv := ⟨h.hom⟩
+
+end add_equiv
+
+
+namespace units
+
+variables [monoid α] [monoid β] [monoid γ]
+(f : α → β) (g : β → γ) [is_monoid_hom f] [is_monoid_hom g]
+
+def map_equiv (h : α ≃* β) : units α ≃* units β :=
+{ to_fun := map h.to_equiv,
+  inv_fun := map h.symm.to_equiv,
+  left_inv := λ u, ext $ h.left_inv u,
+  right_inv := λ u, ext $ h.right_inv u,
+  hom := λ a b, units.ext $ by have X := h.hom; exact X} -- ??
+
```

Comment:
This isn't right. I tried 
```lean
def is_mul_hom {α β : Type*} [has_mul α] [has_mul β] (f : α → β) : Prop :=
∀ ⦃x y⦄, f (x * y) = f x * f y
```

but it seemed to make things worse. What am I doing wrong?

Reply:
I've gone back to round brackets and the problem has disappeared. However I am now tempted to make this a class! So more fiddling!

Reply:
Now a class.